Create Load Default Option for Field and Robot with Gizmo and Assembl…

RoushilS opened the pull request:
AARD-1972

### Description
Added a third button to initial Welcome Modal named Load Default that does the following

1. Loads in the field in cache with field.id equals "1750390764392" (FTC Field 2018 v13).  

2. Loads in the robot in cache with robot.id equals "1750914383413" with Gizmo and Assembly Panel (Dozer v9).   

Other button functionality remains the same.  

### New Welcome Modal:

![Screenshot 2025-06-26 160756](https://github.com/user-attachments/assets/c884ce1f-6601-4af6-8762-c8f523f3df28)

### Upon Clicking Load Default

![image](https://github.com/user-attachments/assets/f7177bf6-c5be-4176-82e0-5cde00b19644)

RoushilS committed:
Create Load Default Option for Field and Robot with Gizmo and Assembly Panel attatched

ryanzhangofficial reviewed (approved):
Lgtm. The formatting issues from before seem to resolved as well 🙌

AlexD717 reviewed (changes requested):
This approach only works if the user has previously cached the robot and field files. On a user's first visit to the website, GetCacheInfo() returns an empty array, so nothing is spawned.

Additionally, each time an object is downloaded from the server, it is assigned a new ID. As a result, the id field cannot be reliably used to identify which field or robot should be spawned.

ryanzhangofficial reviewed (changes requested):
Ah I see that issue too now. I would look at the `manifest` in `ImportMirabufPanel.tsx` for a cleaner solution.

RoushilS committed:
Merge branch 'Autodesk:dev' into dev

RoushilS commented:
Ok, thanks for letting me know, @AlexD717 @ryanzhangofficial 
My understanding is that the default robot and field data get loaded from the manifest, then get sent to the User's cache. They can be retrieved from the cache and spawned. However, the problem I'm finding is that the manifest object is located within ```ImportMirabufPanel```, and there's currently only functionality for the User to cache the default fields and robots. 
I am looking for a different method.

For the identification issue, is searching by name valid? This is from ```/api/mira/manifest.json```
![Screenshot 2025-07-03 134451](https://github.com/user-attachments/assets/02e63b52-4932-4b86-998f-ba6bb001e429)

AlexD717 commented:
I would start by looking at the CacheRemote function in MirabufLoader.ts, I think that fetches the data from the remote. As for filtering by name, that should work.

RoushilS committed:
Fix and New Access Strategy

Initially, was only checking cache. Added step to load from manifest beforehand, then retrieve and spawn.

RoushilS committed:
Format Fix

RoushilS committed:
Remove Unused Imports

AlexD717 reviewed (changes requested):
Could you spawn the 2023 field instead of the 2018 field? Currently the robot gets spawned inside the center structure and is stuck there, while the 2023 field has no center object there.

RoushilS committed:
Use 2023 Field

Uses 2023 field instead of 2018 field to avoid getting stuck within the center structure

rutmanz commented on fission/src/ui/modals/MainMenuModal.tsx:
```diff
@@ -30,6 +33,29 @@ const MainMenuModal: React.FC<ModalPropsImpl & { startSingleplayerCallback: () =
+                    onClick={() => {
+                        closeModal()
+                        startSingleplayerCallback()
+                        MirabufCachingService.CacheRemote(
+                            "/api/mira/fields/FRC Field 2023_v7.mira",
+                            MiraType.FIELD
+                        ).then(cacheInfoField => {
+                            if (cacheInfoField) {
+                                SpawnCachedMira(cacheInfoField, MiraType.FIELD)
+                            }
+                        })
+                        MirabufCachingService.CacheRemote("/api/mira/robots/Dozer_v9.mira", MiraType.ROBOT).then(
+                            cacheInfoRobot => {
+                                if (cacheInfoRobot) {
+                                    SpawnCachedMira(cacheInfoRobot, MiraType.ROBOT)
+                                }
+                            }
+                        )
```
The way these promises are set up means that the order isn't guaranteed and if the robot is spawned first, you have to configure/position it without seeing the field. I think it makes more sense to make sure the field spawns first. You can use `Promise.all(...)` here to do this and still let you download them both simultaneously

RoushilS committed:
Order Promises

Add Promise.All([]) which allows for promises to individually complete and callback and then run sequentially.

AlexD717 reviewed (approved):
Looks good to me

rutmanz commented on fission/src/ui/modals/MainMenuModal.tsx:
```diff
@@ -30,6 +33,31 @@ const MainMenuModal: React.FC<ModalPropsImpl & { startSingleplayerCallback: () =
+                                    }
+                                }
+                            ),
+                        ])
```
This still doesn't guarantee order, because you're still spawning the files immediately after their respective promises resolve, rather than spawning synchronously after both cache requests resolve

[MDN `Promise.all()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all)

Dhruv-0-Arora replied:
```suggestion
               onClick={() => {
                        closeModal()
                        startSingleplayerCallback()
                        let cachedField: MirabufCacheInfo | undefined
                        let cachedRobot: MirabufCacheInfo | undefined
                        Promise.all([
                            MirabufCachingService.CacheRemote(
                                "/api/mira/fields/FRC Field 2023_v7.mira",
                                MiraType.FIELD
                            ).then(cacheInfoField => {
                                if (cacheInfoField) { cachedField = cacheInfoField }
                            }),
                            MirabufCachingService.CacheRemote("/api/mira/robots/Dozer_v9.mira", MiraType.ROBOT).then(
                                cacheInfoRobot => { if (cacheInfoRobot) { cachedRobot = cacheInfoRobot } }
                            ),
                        ]).then(() => {
                            if (cachedField && cachedRobot) {
                                SpawnCachedMira(cachedField, MiraType.FIELD)
                                SpawnCachedMira(cachedRobot, MiraType.ROBOT)
                            } 
                        })
                    }}
```
```suggestion
                    onClick={() => {
                        closeModal()
                        startSingleplayerCallback()
                        Promise.all([
                            MirabufCachingService.CacheRemote(
                                "/api/mira/fields/FRC Field 2023_v7.mira",
                                MiraType.FIELD
                            ).then(cacheInfoField => {
                                if (cacheInfoField) {
                                    SpawnCachedMira(cacheInfoField, MiraType.FIELD)
                                }
                            }),
                            MirabufCachingService.CacheRemote("/api/mira/robots/Dozer_v9.mira", MiraType.ROBOT).then(
                                cacheInfoRobot => {
                                    if (cacheInfoRobot) {
                                        SpawnCachedMira(cacheInfoRobot, MiraType.ROBOT)
                                    }
                                }
                            ),
                        ])
```

Dhruv-0-Arora commented on fission/src/ui/modals/MainMenuModal.tsx:
```diff
@@ -30,6 +33,31 @@ const MainMenuModal: React.FC<ModalPropsImpl & { startSingleplayerCallback: () =
                 />
+                <Button
+                    value={"Load Default"}
+                    onClick={() => {
+                        closeModal()
+                        startSingleplayerCallback()
+                        Promise.all([
+                            MirabufCachingService.CacheRemote(
+                                "/api/mira/fields/FRC Field 2023_v7.mira",
+                                MiraType.FIELD
+                            ).then(cacheInfoField => {
+                                if (cacheInfoField) {
+                                    SpawnCachedMira(cacheInfoField, MiraType.FIELD)
+                                }
+                            }),
+                            MirabufCachingService.CacheRemote("/api/mira/robots/Dozer_v9.mira", MiraType.ROBOT).then(
+                                cacheInfoRobot => {
+                                    if (cacheInfoRobot) {
+                                        SpawnCachedMira(cacheInfoRobot, MiraType.ROBOT)
+                                    }
+                                }
+                            ),
+                        ])
```
```suggestion
                   onClick={() => {
                        closeModal()
                        startSingleplayerCallback()
                        let cachedField: MirabufCacheInfo | undefined
                        let cachedRobot: MirabufCacheInfo | undefined
                        Promise.all([
                            MirabufCachingService.CacheRemote(
                                "/api/mira/fields/FRC Field 2023_v7.mira",
                                MiraType.FIELD
                            ).then(cacheInfoField => {
                                if (cacheInfoField) { cachedField = cacheInfoField }
                            }),
                            MirabufCachingService.CacheRemote("/api/mira/robots/Dozer_v9.mira", MiraType.ROBOT).then(
                                cacheInfoRobot => { if (cacheInfoRobot) { cachedRobot = cacheInfoRobot } }
                            ),
                        ]).then(() => {
                            if (cachedField && cachedRobot) {
                                SpawnCachedMira(cachedField, MiraType.FIELD)
                                SpawnCachedMira(cachedRobot, MiraType.ROBOT)
                            } 
                        })
                    }}
```

Dhruv-0-Arora replied:
This is a solution you could look into for resolving zach's comment

rutmanz replied:
the promise.all Promise returns the individual promise results as an array, you don't need to have `cachedField` and `cachedRobot`

RoushilS committed:
Fixed Promise Ordering: Download in parallel and then spawn in sequence.

rutmanz reviewed (approved):
has some merge conflicts but otherwise looks good

ryanzhangofficial reviewed (commented):
yeah lets get this in🚀

ryanzhangofficial reviewed (approved):
yeah lets get this in🚀

RoushilS committed:
Merge remote-tracking branch 'upstream/dev' into dev

PepperLola committed:
Merge branch 'dev' into dev

PepperLola committed:
fix: added missing imports

BrandonPacewic committed:
fix(format): remaining format issues

BrandonPacewic reviewed (commented):
Awesome work! And congratulations or your first Synthesis commit! 🎊
